src/services/anchor-service.ts

```diff
@@ -156,10 +156,10 @@ export class AnchorService {
     const merkleTree = await this._buildMerkleTree(candidates)
 
     // create and send ETH transaction
-    logger.debug('Preparing to send transaction to put merkle root on blockchain')
-    const tx: Transaction = await this.blockchainService.sendTransaction(
-      merkleTree.getRoot().data.cid
-    )
+    const tx: Transaction = await this.requestRepository.withTransactionMutex(() => {
```

Comment:
what will the behavior wind up being if we keep failing to acquire the mutex forever?

Reply:
As is, this will keep retrying indefinitely to acquire the mutex.

This scenario should not caused be cause by `this.blockchainService.sendTransaction` because this should timeout after 3 attempts. 

But it can be caused by other CAS's always acquiring the mutex. I hope this doesn't happen. This can be helped via some queueing system/nonce management which we can think about for another PR. 

I have built in max attempts and delays which we could configure. I also could add a timeout if needed. But that is uncertain right now. In this scenario the batch will fail. Do you have any thoughts about this or I can bring this up in a discussion at a later time. 

What I can do in the meantime is add alerts every time a CAS cannot acquire a mutex after >5 (arbitrary number) attempts. That way we can know how big of a problem this is.

Reply:
yeah, could also happen if somehow the mutex gets stuck in a locked state.

I see that this process will keep retrying forever, I was wondering more from the perspective of the overall flow.  Is there some higher-level timeout in AWS that will fire eventually?  Or the process keep running forever?  Will we notice if it keeps running forever?  Will this trigger an alert?  What happens when the scheduler service fires off another anchor worker?

Reply:
The mutex should not get stuck locked forever unless something is wrong with postgres or typeorm (more likely). The mutex gets autoreleased at the end of the DB transaction. Also if a session get dropped all of it's locks are released as well. 

I will ask PLAT about these questions. If another anchor worker is launched while one is waiting for the mutex forever, it too will also wait for the lock forever until someone notices which isn't great